File: integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/ObjectWithEmbeddableId.java
package io.quarkus.it.panache;

import java.io.Serializable;
import java.util.Objects;

import jakarta.persistence.Embeddable;
import jakarta.persistence.EmbeddedId;
import jakarta.persistence.Entity;

import io.quarkus.hibernate.orm.panache.PanacheEntityBase;

@Entity
public class ObjectWithEmbeddableId extends PanacheEntityBase {
    @EmbeddedId
    public ObjectKey key;
    public String description;

    @Embeddable
    static class ObjectKey implements Serializable {
        private String part1;
        private String part2;

        public ObjectKey() {
        }

        public ObjectKey(String part1, String part2) {
            this.part1 = part1;
            this.part2 = part2;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o)
                return true;
            if (o == null || getClass() != o.getClass())
                return false;
            ObjectKey objectKey = (ObjectKey) o;
            return part1.equals(objectKey.part1) &&
                    part2.equals(objectKey.part2);
        }

        @Override
        public int hashCode() {
            return Objects.hash(part1, part2);
        }
    }
}

File: integration-tests/hibernate-reactive-panache-kotlin/src/main/kotlin/io/quarkus/it/panache/reactive/kotlin/ObjectWithEmbeddableId.kt
package io.quarkus.it.panache.reactive.kotlin

import io.quarkus.hibernate.reactive.panache.kotlin.PanacheCompanionBase
import io.quarkus.hibernate.reactive.panache.kotlin.PanacheEntityBase
import jakarta.persistence.Embeddable
import jakarta.persistence.EmbeddedId
import jakarta.persistence.Entity
import java.io.Serializable
import java.util.Objects

@Entity
class ObjectWithEmbeddableId : PanacheEntityBase {
    companion object : PanacheCompanionBase<ObjectWithEmbeddableId, ObjectKey>

    @EmbeddedId
    var key: ObjectKey? = null
    var description: String? = null

    @Embeddable
    class ObjectKey : Serializable {
        private var part1: String? = null
        private var part2: String? = null

        constructor() {}
        constructor(part1: String?, part2: String?) {
            this.part1 = part1
            this.part2 = part2
        }

        override fun equals(o: Any?): Boolean {
            if (this === o) return true
            if (o == null || javaClass != o.javaClass) return false
            val objectKey = o as ObjectKey
            return part1 == objectKey.part1 && part2 == objectKey.part2
        }

        override fun hashCode(): Int {
            return Objects.hash(part1, part2)
        }
    }
}
